File: problem_14.cpp
#include <chrono>
#include <iostream>
#include <array>
#include <string>
#include <cstring>
#include <cmath>

const int maxStarter = 1000000;

long nextCollatz (long i) 
{
    switch (i % 2)
    {
        case 0:
        return i / 2;
        break;
        case 1:
        default:
        return 3 * i + 1;
    }
}


int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();

    std::array<int, maxStarter> accCollatz;
    int* pCollatz = accCollatz.data();
    memset(pCollatz, 0, maxStarter*sizeof(int));
    int longestSequence = 0;
    int longestSequenceNum = 0;
    int filledAccCollatz = 1;

    for (long i = 1; i < maxStarter && filledAccCollatz < maxStarter; i++) 
    {
        if (pCollatz[i])
        {
            continue;
        }
        pCollatz[i] = 1;
        filledAccCollatz++;
        long cur = i;
        int curSequence = 0;
        while (cur != 1)
        {
            cur = nextCollatz(cur);
            curSequence++;
            if (cur > 1 && cur < maxStarter && !pCollatz[i])
            {
                pCollatz[i] = 1;
                filledAccCollatz++;
            }
        }
        if (curSequence > longestSequence) 
        {
            longestSequence = curSequence;
            longestSequenceNum = i; 
        }
    }

    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Sequence length: " << longestSequence << std::endl;
    std::cout << "Solution: " << longestSequenceNum << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_14.rs
use std::time::{Instant};

const MAX_STARTER: usize = 1000000;

fn next_collatz (i: usize) -> usize
{
    return match i % 2
    {
        0 => i / 2,
        1 => 3 * i + 1,
        _ => 1,
    }
}

fn main () 
{
    let start = Instant::now();

    let mut acc_collatz: [bool; MAX_STARTER] = [false; MAX_STARTER];
    let mut longest_sequence: usize = 0;
    let mut longest_sequence_num: usize = 0;
    let mut filled_acc_collatz: usize = 1;

    for i in 1..MAX_STARTER
    {
        if acc_collatz[i]
        {
            continue;
        }
        acc_collatz[i] = true;
        filled_acc_collatz = filled_acc_collatz + 1;
        let mut cur: usize = i;
        let mut cur_sequence: usize = 0;
        while cur != 1
        {
            cur = next_collatz(cur);
            cur_sequence = cur_sequence + 1;
            if cur > 1 && cur < MAX_STARTER && !acc_collatz[i]
            {
                acc_collatz[i] = true;
                filled_acc_collatz = filled_acc_collatz + 1;
            }
        }
        if filled_acc_collatz >= MAX_STARTER
        {
            break;
        }
        if cur_sequence > longest_sequence
        {
            longest_sequence = cur_sequence;
            longest_sequence_num = i; 
        }
    }

    println!("Result: {}", longest_sequence_num);
    let duration = start.elapsed();
    println!("Timing: {}", duration.as_nanos());
}



